torch/_inductor/scheduler.py

```diff
@@ -2416,21 +2401,38 @@ def can_fuse_vertical(
                 why("intermediate nodes between node1 & node2")
                 return False
 
-        # similar to can_inplace, if we are going to fuse a write subsequent to a read
-        # require that the indexing and size is the same
-        for write in node2.read_writes.writes:
-            if not isinstance(write, MemoryDep):
-                continue
-            for read in node1.read_writes.reads:
-                if write.name != self.mutation_renames.get(read.name, read.name):
-                    continue
+        return True
 
-                # bail on StarDep
-                if not self.fusable_read_and_write(read, write):
-                    why("fusing a write into a read with different indexing formula")
-                    return False
+    def fusable_weak_dep(
+        self, weak_dep: WeakDep, node1: BaseSchedulerNode, node2: BaseSchedulerNode
+    ) -> bool:
+        # A weak dep can be fused if and only if the fused operation acts inplace
+        # on the buffer being mutated. i.e. the same index is being read then mutated
+        mutating_writes = [
+            write
+            for write in node2.read_writes.writes
+            if write.name == weak_dep.mutating_buf
+        ]
+        if len(mutating_writes) != 1:
+            return False
+        write = mutating_writes[0]
+        assert isinstance(write, MemoryDep)
 
-        return True
+        if free_symbol_is_type(write.index, SymT.TMP):
+            return False
```

Comment:
What if the read and the write use the same indirect indexing? Isn't it enough to check that `read.index == write.index` and same for the size?

Reply:
The index might contain a value `tmp0` which could have different values from one kernel to the next, so it's not possible to compare index expressions with tmp variables between kernels.

Reply:
and is it not possible to compare the submodules that compute that indirect index? If these are equal, then we could still fuse them.

Reply:
The read and write are always from two different nodes here.

Reply:
Or I guess you mean doing an inspection of the ir graph that produced the indirect_index call. I guess it should be possible but it's out of scope for this change.

Reply:
Yep, that's what I meant. And yeah, if we do that, we should probably first implement it in the general fusion algorithm.